File: 2/part1.js
const _ = require('lodash');
const passwords = require('./passwords.json');

const validPasswords = passwords.filter(line => {
  const {min, max, char, password} = parseLine(line);
  const fil = _.filter(password, c => c === char);
  const count = fil.length;
  let valid = count >= min && count <= max;
  return valid;
});

console.log(validPasswords.length);

function parseLine(line) {
  const split = line.split(' ');
  const minmax = split[0].split('-');
  const min = parseInt(minmax[0]);
  const max = parseInt(minmax[1]);
  const char = split[1][0];
  const password = split[2];

  return {min, max, char, password};
}

File: 2/hs/part1.hs
import qualified Data.Text as Text
import qualified Data.Text.IO as Text
import Parselib

type PasswordPolicy = (Int, Int, Char, String)

line :: Parser PasswordPolicy
line = do
  min <- natural
  symb "-"
  max <- natural
  space
  c <- letter
  symb ":"
  password <- many letter
  space
  return (min, max, c, password)

parseLine :: String -> PasswordPolicy
parseLine l = case apply line l of
  [(a, "")] -> a
  _ -> error l

valid :: String -> Bool
valid l = do
  let (min, max, char, password) = parseLine l
  let fil = filter (== char) password
  let count = length fil
  min <= count && count <= max

main = do
  text <- Text.readFile "passwords.txt"
  let list = map Text.unpack $ Text.lines text
  let fil = filter valid list
  let count = length fil
  print count
